plugins/cpp/parser/src/cppparser.cpp

```diff
@@ -339,7 +343,86 @@ std::vector<std::string> CppParser::getDependentParsers() const
   return std::vector<std::string>{};
 }
 
-bool CppParser::preparse(bool dry_)
+std::vector<std::vector<std::string>> CppParser::createCleanupOrder()
```

Comment:
What is the use of this ordering? Isn't it enough simply removing all records in all tables belonging to the modified/deleted files?

Reply:
As described in this PR's description:

> Cleanup is performed in a parallel manner. For this to be achieved, a graph is constructed from the files to be cleaned up along the inclusion relations. Then the nodes of the graph are sorted topologically. In each iteration the leaf nodes of the graph are cleaned up (paralelly), which guarantees that no file is cleaned up before its dependants.

So, if e.g. `a.cpp` depends on `b.h` and we clean up `b.h` before `a.cpp`, that would fail the cleanup process.

Reply:
To reproduce the error that made topological ordering necessary, let's say we have these files:

definitions.h:
```
#ifndef DEFINITIONS_H
#define DEFINITIONS_H

enum class Day
{
  Monday = 1,
  Tuesday,
  Wednesday,
  Thursday,
  Friday,
  Saturday,
  Sunday  
};

enum Permission
{
  Readable = 4,
  Writable = 2,
  Executable = 1
};

#endif // DEFINITIONS_H
```

main.cpp:
```
#include <iostream>

#include "definitions.h"

int main()
{
  Day d = Day::Friday;
  std::cout << "d = " << static_cast<int>(d) << std::endl;
  d = Day::Sunday;
  std::cout << "d = " << static_cast<int>(d) << std::endl;
  
  int p = Readable;
  std::cout << "p = " << p << std::endl;
  p |= Executable;  
  std::cout << "p = " << p << std::endl;
  
  return 0;
}
```

Any modification in _definitions.h_ produces the following error when the arbitrarily ordered cleanup is executed:

> [INFO] (1/1) Parsing /home/efekane/Repos/cc-tests/enums/main.cpp
[DEBUG] CppMacro
id = 0
mangledNameHash = 13147581534377510727
qualifiedName = DEFINITIONS_H
[ERROR] 23505: ERROR:  duplicate key value violates unique constraint "CppEntity_astNodeId_i"
DETAIL:  Key ("astNodeId")=(3956703780563145061) already exists.

Reply:
I still couldn't reproduce this error. I created a branch "no_order" on my fork where I appended a commit in which the topological ordering is skipped. Could you please checkout and run that code? If you can reproduce the error, could you please describe a scenario? Thanks!

Reply:
This error cannot be reproduced and might be a result of the usage of an outdated codebase or workspace database. We can drop this issue.
However we still have issues related to the parallel cleanup of files with include dependency (which can be prevented with the topological sorting), I emailed you the details @bruntib .